content/en/blog/_posts/2025-01-14-devices-failure-handling/index.md

```diff
@@ -0,0 +1,403 @@
+{{< /table >}}
+
+The existing failure model was relying on old assumptions. It may still work for
+the new workload types, but it has limited knowledge about devices and is very
+expensive for them. In some cases, even prohibitively expensive. You will see
+more examples later in this article.
+
+### Why Kubernetes still reigns supreme
+
+This article is not going deeper into the question of "why not start fresh" for
+AI/ML workloads since they are so different from the traditional Kubernetes
+workloads. Despite many challenges, Kubernetes remains the platform of choice
+for AI/ML workloads. Its maturity, security, and rich ecosystem of tools make it
+a compelling option. While alternatives exist, they often lack the years of
+development and refinement that Kubernetes offers. And the Kubernetes developers
+are actively addressing the gaps identified in this article and beyond.
+
+## The current state of device failure handling
+
+This section outlines different failure modes and the best practices and DIY
+(Do-It-Yourself) solutions used today. The next session will describe a roadmap
+of improving things for those failure modes.
+
+### Failure modes: k8s infra
+
+In order to understand the failures related to the Kubernetes infrastructure,
+you need to understand how many moving parts are involved in scheduling a Pod on
+the node. The sequence of events when the Pod is scheduled in the Node is as
+follows:
+
+1. *Device plugin* is scheduled on the Node  
+1. *Device plugin* is registered with the *kubelet* via local gRPC  
+1. *Kubelet* uses *device plugin* to watch for devices and updates capacity of
+   the node  
+1. *Scheduler* places a *User Pod* on a Node based on the updated capacity  
+1. *Kubelet* asks *Device plugin* to **Allocate** devices for a *User Pod*  
+1. *Kubelet* creates a *User Pod* with the allocated devices attached to it
+
+This diagram shows some of those actors involved:
+
+{{< figure src="k8s-infra-devices.svg" alt="General diagram of how device plugin works" >}}
+
+As there are so many actors interconnected, every one of them and every
+connection may experience interruptions. This leads to many exceptional
+situations that are often considered failures, and may cause serious workload
+interruptions:
+
+* Pods failing admission at various stages of its lifecycle  
+* Pods unable to run on perfectly fine hardware  
+* Scheduling taking unusually long time
+
+{{< figure src="k8s-infra-failures.svg" alt="General diagram of possible failures in how device plugin works" >}}
+
+The goal for Kubernetes is to make the interruption between these components as
+reliable as possible. Kubelet already implements retries, grace periods, and
+other techniques to improve it. The roadmap section goes into details on other
+edge cases that the Kubernetes project tracks. However, all these improvements
+only work when these best practices are followed:
+
+* Configure and restart kubelet and the container runtime (containerd or CRI-O)
```

Comment:
```suggestion
* Configure and restart kubelet and the container runtime (such as containerd or CRI-O)
```

We don't consider those two runtimes a best practice; using another compatible runtime is fine. We do _test_ against those two (only) but that does not mean the same thing.